File: 92_reverse_linked_list_II/reverseBetween.py
class Solution(object):
    def reverseList(self, head):
        prev = None
        while head:
            curr = head
            head = head.next
            curr.next = prev
            prev = curr
        
        return prev

    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """

        tmp = head
        n -= m
        
        if n == 0: return head

        prev = None
        
        while m > 1:
            prev = tmp
            tmp = tmp.next
            m -= 1
       
        begin = tmp

        #if not tmp:
        #    return head
        
        while n > 0:
            tmp = tmp.next
            n -= 1
        
        end_next = tmp.next
        tmp.next = None
        rev_list = self.reverseList(begin)
        
        if prev:
            prev.next = rev_list
        else:
            head = rev_list
        
        tmp = rev_list
        while tmp.next:
            tmp = tmp.next

        tmp.next = end_next

        return head

File: 92_reverse_linked_list_II/reverseBetween.cpp
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


class Solution {
	public:
    	ListNode* reverseBetween(ListNode* head, int m, int n) {
    	    ListNode* go = head;
            while(--m) {
                go = go -> next;
                n--;
            }
            ListNode* rHead = go;
            while(--n) {
                go = go -> next;
            }
            ListNode* rTail = go;
            reversePart(rHead, rTail, rTail -> next);
            return head;
        }
    private:
        ListNode* reversePart(ListNode* head, ListNode* tail, ListNode* tailNext) {
            ListNode* prev = tailNext;
            ListNode* curr;
            while(head != tail) {
                curr = head;
                head = head -> next;
                curr -> next = prev;
                prev = curr;
            }
            return prev;
        }
};

int main() {
    return 0;
}
